core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1695,30 +1720,64 @@ object KafkaZkClient {
+        case _ => throw new IllegalStateException(s"$request does not need controller epoch check")
+      }
+  }
 
-  private def controllerZkVersionCheck(version: Int): Option[ZkVersionCheck] = {
-    if (version < KafkaController.InitialControllerEpochZkVersion)
-      None
-    else
-      Some(ZkVersionCheck(ControllerEpochZNode.path, version))
-  }
-
-  private def maybeThrowControllerMoveException(response: AsyncResponse): Unit = {
-    response.zkVersionCheckResult match {
-      case Some(zkVersionCheckResult) =>
-        val zkVersionCheck = zkVersionCheckResult.zkVersionCheck
-        if (zkVersionCheck.checkPath.equals(ControllerEpochZNode.path))
-          zkVersionCheckResult.opResult match {
-            case errorResult: ErrorResult =>
-              val errorCode = Code.get(errorResult.getErr)
-              if (errorCode == Code.BADVERSION)
-              // Throw ControllerMovedException when the zkVersionCheck is performed on the controller epoch znode and the check fails
-                throw new ControllerMovedException(s"Controller epoch zkVersion check fails. Expected zkVersion = ${zkVersionCheck.expectedZkVersion}")
-              else if (errorCode != Code.OK)
-                throw KeeperException.create(errorCode, zkVersionCheck.checkPath)
-            case _ =>
-          }
-      case None =>
+  // A helper function to transform a MultiResponse with the check on
+  // controller epoch znode zkVersion back into a regular response.
+  // ControllerMovedException will be thrown if the controller epoch
+  // znode zkVersion check fails. This is used for fencing zookeeper
+  // updates in controller.
+  private def unwrapResponseWithControllerEpochCheck(response: AsyncResponse): AsyncResponse = {
+    response match {
+      case MultiResponse(resultCode, _, ctx, zkOpResults, responseMetadata) =>
+        zkOpResults match {
+          case Seq(ZkOpResult(checkOp: CheckOp, checkOpResult), zkOpResult) =>
+            checkOpResult match {
+              case errorResult: ErrorResult =>
+                if (checkOp.path.equals(ControllerEpochZNode.path)) {
+                  val errorCode = Code.get(errorResult.getErr)
+                  if (errorCode == Code.BADVERSION)
+                  // Throw ControllerMovedException when the zkVersionCheck is performed on the controller epoch znode and the check fails
+                    throw new ControllerMovedException(s"Controller epoch zkVersion check fails. Expected zkVersion = ${checkOp.version}")
+                  else if (errorCode != Code.OK)
+                    throw KeeperException.create(errorCode, checkOp.path)
+                }
+              case _ =>
+            }
+            val rawOpResult = zkOpResult.rawOpResult
+            zkOpResult.zkOp match {
+              case createOp: CreateOp =>
+                val name = rawOpResult match {
+                  case c: CreateResult => c.getPath
```

Comment:
According to the zookeeper client Javadoc, the name in `CreateResponse` is expected to be `The name of the Znode that was created`. Also, it is mentioned that `On success, <i>name</i> and <i>path</i> are usually, equal, unless a sequential node has been created`.

On the other hand, the Javadoc for `CreateResult` says that, `A result from a create operation.  This kind of result allows the path to be retrieved since the create might have been a sequential create`.

We need to make sure that the `CreateResult.path` has the same value as the original value of the `name` in `CreateResponse` when a sequential node is created. The Javadoc  `CreateResult` suggests this is the case but the name of its variable, i.e. `path`, suggests they are different. Can you double check this by creating a sequential znode?